R/cellbrowser.R

```diff
@@ -0,0 +1,499 @@
+  }
+  if (substr(x = object.version, start = 1, stop = 1) != substr(x = packageVersion(pkg = "Seurat"), start = 1, stop = 1)) {
+    stop("The installed version of Seurat is different from Seurat object loaded. You have to down- or upgrade your installed Seurat version, see the Seurat documentation")
+  }
+  embeddings <- reductions # Seurat prefers the word reductions
+  # compatibility layer for Seurat 2 vs 3
+  # see https://satijalab.org/seurat/essential_commands.html
+  if (substr(x = object.version, start = 1, stop = 1) == '2') {
+    # Seurat 2 data access
+    idents <- object@ident # Idents() in Seurat3
+    meta <- object@meta.data
+    cellOrder <- object@cell.names
+    if (matrix.slot == "counts") {
+      counts <- object@raw.data
+    } else if (matrix.slot == "scale.data") {
+      counts <- object@scale.data
+    }
+    else if (matrix.slot == "data") {
+      counts <- object@data
+    } else {
+      stop("matrix.slot can only be one of: counts, scale.data, data")
+    }
+    genes <- rownames(x = object@data)
+    dr <- object@dr
+  } else {
+    # Seurat 3 functions
+    idents <- Idents(object = object)
+    meta <- object[[]]
+    cellOrder <- colnames(x = object)
+    counts <- findMatrix(object = object, matrix.slot = matrix.slot)
+    if (dim(x = counts)[1] == 0) {
+      message(paste0("The Seurat data slot '", matrix.slot, "' contains no data. Trying default assay."))
+      defAssay <- DefaultAssay(object)
+      assay <- GetAssay(object, defAssay)
+      message(paste0("Default assay is ", defAssay))
+      counts <- findMatrix(assay, matrix.slot)
+      genes <- rownames(counts)
+      if (dim(x = counts)[1] == 0) {
+        stop(
+          "Could not find an expression matrix",
+          "Please select the correct slot where the matrix is stored, possible ",
+          "values are 'counts', 'scale.data' or 'data'. To select a slot, ",
+          "use the option 'matrix.slot' from R or the cbImportSeurat option -s from the command line."
+        )
+      }
+    }
+    else {
+      genes <- rownames(x = object)
+    }
+    dr <- object@reductions
+  }
+  if (is.null(x = cluster.field)) {
+    cluster.field = "Cluster"
+  }
+  if (is.null(x = meta.fields)) {
+    meta.fields <- colnames(x = meta)
+    if (length(x = levels(x = idents)) > 1) {
+      meta.fields <- c(meta.fields, ".ident")
+    }
+  }
+  if (!is.null(x = port) && is.null(x = cb.dir)) {
+    stop("cb.dir parameter is needed when port is set")
+  }
+  if (!dir.exists(paths = dir)) {
+    dir.create(path = dir)
+  }
+  if (!dir.exists(paths = dir)) {
+    stop("Output directory ", dir, " cannot be created or is a file")
+  }
+  if (dataset.name == "SeuratProject") {
+    warning("Using default project name means that you may overwrite project with the same name in the cellbrowser html output folder")
+  }
+  enum.fields <- c()
+  # Export expression matrix
+  if (!skip.expr.matrix) {
+    if (use.mtx) {
+      require(Matrix)
+      require(R.utils)
```

Comment:
CRAN dependencies should be declared with `@importFrom` and listed in the `Imports:` section of DESCRIPTION, not called with `require`

Reply:
done.